### PR TITLE
Implement editing of domain bounds

### PR DESCRIPTION
**Pull request by axelboc**
Some screenshots for 😮, 😋 and 😍 ... (at least! 😏)

image

image

image

image

image

### CODE REVIEW

File: src/h5web/toolbar/Toolbar.module.css
Comment (axelboc) [inline src/h5web/toolbar/Toolbar.module.css:37]: Moved to global `btn-clean`.

File: src/h5web/toolbar/Toolbar.module.css
Comment (axelboc) [inline src/h5web/toolbar/Toolbar.module.css:88]: Buttons now look pressed on `mousedown`.

File: src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx
Comment (axelboc) [inline src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx:66]: This "pins" the popup open when a field receives focus.

File: src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx
Comment (axelboc) [inline src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:35]: The slider domain state now holds the "unsafe" vis domain. This allows the tooltip to show the custom domain even when it's invalid, while also showing the slider's domain while a thumb is being moved.

The `safeDomain` is passed to the `ScaledSlider` for computing the extended domain. `ScaledSlider` then recomputes a safe domain from its `value` on every render so the thumbs don't end up in weird places.

File: src/h5web/vis-packs/core/heatmap/utils.ts
Comment (loichuder) [inline src/h5web/vis-packs/core/heatmap/utils.ts:25]: `dataDomain` could be renamed to `fallbackDomain`

File: src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx
Comment (loichuder) [inline src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:109]: Not sure about this. It feels odd to toggle editing when deactivating the autoscale.

My expectation would be that the focus on the edition is lost when toggling a button (i.e. `toggleEditingMin(false)`).

File: src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx
Comment (loichuder) [inline src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:92]: I feel the icon is not really appropriate here as it shows sliders.

Perhaps `MdEdit` or `FiEdit` would be more manifest ?

File: src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx
Comment (loichuder) [inline src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx:61]: Isn't it rather https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/issues/566 ?

File: src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx
Comment (loichuder) [inline src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx:51]: I find counter-intuitive to have the `value` on the label.

Isn't there a way to have it on hover on the editing box when not editing ?

File: src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx
Comment (loichuder) [inline src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx:24]: I would find more intuitive if the value was staying under `.3e` format when editing.

But that's my personal opinion so let's wait for user feedback before changing.

**Review by loichuder — CHANGES_REQUESTED** [review]
I like the design very much ! Looking clean !

A few nitpicks here and there

File: src/h5web/vis-packs/core/heatmap/utils.ts
Comment (axelboc) [reply src/h5web/vis-packs/core/heatmap/utils.ts:25]: and `visDomain` to just `domain`

File: src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx
Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:109]: Oh yeah? I thought maybe users who disable the autoscale would want to edit the value right away in some cases. That being said, it does "pin" the popup, which means an extra click is required to close it if they're not interested in editing the value... Alright, I'll change this.

Comment (loichuder) [reply src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:109]: > That being said, it does "pin" the popup, which means an extra click is required to close it if they're not interested in editing the value... Alright, I'll change this.

Yup that was my point

Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:109]: However, I think if a field is in edit mode and the user disables the autoscale, I think the field should remain in edit mode.

File: src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx
Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:92]: Yeah, but those are really ugly on their own... Do you feel strongly about it? The icon is not meant to describe the UI but the fact that the domain can be "tuned". As long as users understand that... 🤷

File: src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx
Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx:61]: haha yes, thanks, forgot to update the link 😄

File: src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx
Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx:51]: I'll give it a try.

File: src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx
Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx:24]: Ah yeah? It would allow me to shrink back the size of the input fields and to not have to reduce the font-size in editing mode. It wouldn't prevent people from entering values with as many decimals as they want, so... why not!

Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx:24]: > so let's wait for user feedback before changing.

yeah agreed

File: src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx
Comment (loichuder) [reply src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:92]: > Yeah, but those are really ugly on their own...

Really ? I like `FiEdit`
image



> The icon is not meant to describe the UI but the fact that the domain can be "tuned".

I get you but the icon aspect is really at odds with the fact that clicking on it toggles editing. Well in any case:

> Do you feel strongly about it? 

No :smile:

File: src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx
Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/BoundEditor.tsx:51]: Alright, as I thought, I can just use the `title` attribute on the input field. The downside is that there no longer is an indication that supplementary information is available like there was with `abbr` (the dotted underline). Not a huge problem.

image

**Review by loichuder — APPROVED** [review]
Great !

**Comment by loichuder** [comment]
Before you merge, I noticed a small issue:

- Click on the **Tune** button to toggle edit
- Click on it again to toggle edit off

The tooltip disappears and do not appear again, even when moving thumbs. One must move the pointer away from the slider to make it reappear.

File: src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx
Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:92]: `FiEdit` isn't too bad, but it looks wrong visually because the square outline is not centred in the button (and I think tweaking the alignment would require reducing the icon's size too much).

But I'm not giving up yet! Here are a couple of alternatives (note that I can adjust the color/stroke to make some of these less imposing):

image

image

image

image

image

image

... and my new favourite: 

image

**Comment by axelboc** [comment]
> The tooltip disappears and do not appear again, even when moving thumbs. One must move the pointer away from the slider to make it reappear.

Yeah, it's on purpose. You would only click this button to explicitly close the popup and cancel your edits. If you click and the popup stays, it just feels confusing... doesn't it?

File: src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx
Comment (axelboc) [reply src/h5web/toolbar/controls/DomainSlider/DomainSlider.tsx:92]: With the `small` variant applied to the toggle:

image

**Comment by axelboc** [comment]
Neh you're right, now with the input fields, the behaviour of keeping the tooltip open makes more sense.

**Pull request merged**
